WS2812.class.nut

```diff
@@ -2,121 +2,126 @@
-	
-	function _initialize() 
-	{
-		// Fill the bits array first
-
-		_bits = array(256);
-		
-		for (local i = 0; i < 256; i++) 
-		{
-			local valblob = blob(BYTESPERPIXEL / 3);
-			valblob.writen((i & 0x80) ? ONE:ZERO,'b');
-			valblob.writen((i & 0x40) ? ONE:ZERO,'b');
-			valblob.writen((i & 0x20) ? ONE:ZERO,'b');
-			valblob.writen((i & 0x10) ? ONE:ZERO,'b');
-			valblob.writen((i & 0x08) ? ONE:ZERO,'b');
-			valblob.writen((i & 0x04) ? ONE:ZERO,'b');
-			valblob.writen((i & 0x02) ? ONE:ZERO,'b');
-			valblob.writen((i & 0x01) ? ONE:ZERO,'b');
-			_bits[i] = valblob;
-		}
-
-		// Now fill the clearblob
-		
-		_clearblob = blob(BYTESPERPIXEL);
-		
-		for (local j = 0 ; j < BYTESPERPIXEL ; j++) 
-		{
-			_clearblob.writen(ZERO, 'b');
-		}
-	}
+class WS2812 {
+    // This class uses SPI to emulate the WS2812s' one-wire protocol.
+    // This requires one byte per bit to send data at 7.5 MHz via SPI.
+    // These consts define the "waveform" to represent a zero or one
+
+    static version = [2,0,0];
+
+    static ZERO            = 0xC0;
+    static ONE             = 0xF8;
+    static BYTES_PER_PIXEL   = 24;
+
+    // When instantiated, the WS2812 class will fill this array with blobs to
+    // represent the waveforms to send the numbers 0 to 255. This allows the blobs to be
+    // copied in directly, instead of being built for each pixel - which makes the class faster.
+
+    static _bits     = array(256);
+
+    // Private variables passed into the constructor
+
+    _spi             = null;  // imp SPI interface (pre-configured)
+    _frameSize       = null;  // number of pixels per frame
+    _frame           = null;  // a blob to hold the current frame
+
+    // Parameters:
+    //    spi          A pre-configured SPI bus (MSB_FIRST, 7500)
+    //    frameSize    Number of Pixels per frame
+
+    constructor(spiBus, frameSize) {
+        // spiBus must be configured
+        _spi = spiBus;
+
+        _frameSize = frameSize;
+        _frame = blob(_frameSize * BYTES_PER_PIXEL + 1);
+        _frame[_frameSize * BYTES_PER_PIXEL] = 0;
+
+        // Used in constructing the _bits array
+        local bytesPerColor = BYTES_PER_PIXEL / 3;
+
+        // Fill the _bits array if required
+        // (Multiple instance of WS2812 will only initialize it once)
+        if (_bits[0] == null) {
+            for (local i = 0; i < 256; i++) {
+                local valblob = blob(bytesPerColor);
+                valblob.writen((i & 0x80) ? ONE:ZERO,'b');
+                valblob.writen((i & 0x40) ? ONE:ZERO,'b');
+                valblob.writen((i & 0x20) ? ONE:ZERO,'b');
+                valblob.writen((i & 0x10) ? ONE:ZERO,'b');
+                valblob.writen((i & 0x08) ? ONE:ZERO,'b');
+                valblob.writen((i & 0x04) ? ONE:ZERO,'b');
+                valblob.writen((i & 0x02) ? ONE:ZERO,'b');
+                valblob.writen((i & 0x01) ? ONE:ZERO,'b');
+                _bits[i] = valblob;
+            }
+        }
+
```

Comment:
Should we have a constructor parameter which allows you to *not* write zeros to the leds? In cases where you are displaying something, it would allow you to reboot without the display glitching.

constructor(spiBus, frameSize, initialize=true)